submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+## Globalnet 2.0: Enhancements to Globalnet Implementation
+
+## Summary
+
+Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)
+annotates every Pod, Service and Node in the cluster with a globalIp to support Pod to remote
+Service connectivity in the following manner.
+
+Once a Pod is scheduled and enters Running state, it is annotated with a globalIp. To support
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIp. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
+with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIp.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+1. By default, every cluster will be assigned a single globalIp which will be used as egress IP
+   for any cross-cluster communication.
+2. Provision to allocate a globalIp at the namespace/project level which takes precedence over
+   the globalIp allocated to the cluster.
+3. Provision to allocate a globalIp to a Pod (or a set of Pods) which has higher precedence
+   over globalIPs at namespace level. If there are multiple policies, users can specify a priority
```

Comment:
Priority is not really an important feature IMHO but for sure we want consistency, so that a single pod doesnt flip between IPs.
I feel how we do it is better left as implementation details (as user will likely not care).

Reply:
It's something that we have identified in brainstorms with the community.

Reply:
But may be it's worth leaving this for the future, as it's a simple additive improvement, and simplifies the iptables a lot. (Also probably in OVN)

Reply:
Say we remove Priority (atleast for the initial implementation) and user creates multiple GlobalnetEgressIP CRDs where PodSelectors match common Pods. The behavior (ie., globalIP which will be used) cannot be well defined. Would this be acceptable?

CC: @nyechiel

Reply:
@sridhargaddam it could be that one policy matches a namespace, and another policy specific pods.

May be we should put those in separate iptable chains, giving namespace only selectors a first priority chain, and more complex selectors a lower priority.

Then in the future we can think about priority. I suspect this would satisfy most cases.

Reply:
Okay, what about multiple GlobalnetEgressIP objects where PodSelectors match common Pods.

i.e., say we have pods - p1, p2, p3, p4 and p5.
GlobalnetEgressIP object1 - matches p1 and p2
GlobalnetEgressIP object2 - matches p2 and p4

What should be the egress globalIP for p2 which has a match in both the objects? Can we say that it will be implementation-specific?

Reply:
We should say it's not deterministic, and any of both could be expected.

Reply:
Thanks @mangelajo. I had look at OpenShift EgressIP feature and even over there its mentioned that its non deterministic.

"_When a pod matches the selector for multiple EgressIP objects, there is no guarantee which of the egress IP addresses that are specified in the EgressIP objects is assigned as the egress IP address for the pod._"

Reply:
I'm removing the Priority field.